causalpy/pymc_models.py

```diff
@@ -507,3 +507,295 @@ def fit(self, X, t, coords):
                 )
             )
         return self.idata
+
+
+class InterventionTimeEstimator(PyMCModel):
+    r"""
+    Custom PyMC model to estimate the time an intervention took place.
+
+    defines the PyMC model :
+
+    .. math::
+        \beta &\sim \mathrm{Normal}(0, 1) \\
+        \mu &= \beta \cdot X\\
+        \\
+        \tau &\sim \mathrm{Uniform}(0, 1) \\
+        w &= sigmoid(t-\tau) \\
+        \\
+        \text{level} &\sim \mathrm{Normal}(0, 1) \\
+        \text{trend} &\sim \mathrm{Normal}(0, 1) \\
+        A &\sim \mathrm{Normal}(0, 1) \\
+        \lambda &\sim \mathrm{HalfNormal}(0, 1) \\
+        \text{impulse} &= A \cdot exp(-\lambda \cdot |t-\tau|) \\
+        \mu_{in} &= \text{level} + \text{trend} \cdot (t-\tau) + \text{impulse}\\
+        \\
+        \sigma &\sim \mathrm{HalfNormal}(0, 1) \\
+        \mu_{ts} &= \mu + \mu_{in} \\
+        \\
+        y &\sim \mathrm{Normal}(\mu_{ts}, \sigma)
+
+    Example
+        --------
+        >>> import causalpy as cp
+        >>> import numpy as np
+        >>> from patsy import build_design_matrices, dmatrices
+        >>> from causalpy.pymc_models import InterventionTimeEstimator as ITE
+        >>> data = cp.load_data("its")
+        >>> formula="y ~ 1 + t + C(month)"
+        >>> y, X = dmatrices(formula, data)
+        >>> outcome_variable_name = y.design_info.column_names[0]
+        >>> labels = X.design_info.column_names
+        >>> _y, _X = np.asarray(y), np.asarray(X)
+        >>> _X = xr.DataArray(
+        ... _X,
+        ... dims=["obs_ind", "coeffs"],
+        ... coords={
+        ...     "obs_ind": data.index,
+        ...     "coeffs": labels,
+        ...     },
+        ... )
+        >>> _y = xr.DataArray(
+        ...     _y[:, 0],
+        ...     dims=["obs_ind"],
+        ...     coords={"obs_ind": data.index},
+        ...     )
+        >>> COORDS = {"coeffs":labels, "obs_ind": np.arange(_X.shape[0])}
+        >>> model = ITE(treatment_effect_type="level", sample_kwargs={"draws" : 10, "tune":10, "progressbar":False})
+        >>> model.set_time_range(None, data)
+        >>> model.fit(X=_X, y=_y, coords=COORDS)
+        Inference ...
+    """
+
+    def __init__(
+        self,
+        treatment_effect_type: str | list[str],
+        treatment_effect_param=None,
+        sample_kwargs=None,
+    ):
+        """
+        Initializes the InterventionTimeEstimator model.
+
+        :param treatment_effect_type: Optional dictionary that specifies prior parameters for the
+            intervention effects. Expected keys are:
```

Comment:
This is interesting, so similar to change-point detection, but instead of a mean difference in the level you also look for impulses and trend shifts...but why is the trend and impulse parameterised with sigma and mu? That seems like it's mean differences again?